FunNet/Classes/Core/NetworkErrorHandler.swift

```diff
@@ -120,3 +122,330 @@ extension UIAlertController {
+    
+    public init(configs: ErrorHandlingConfig..., errorMessageMap: [Int:String] = [:], serverErrorMessageMap: [Int:String] = [:], presenter: @escaping (UIViewController) -> Void) {
+        self.config = configs.map(\.rawValue).reduce(0, { $0 | $1 })
+        self.presenter = presenter
+        for (code, message) in errorMessageMap {
+            self.errorMessageMap[code] = message
+        }
+        for (code, message) in serverErrorMessageMap {
+            self.serverErrorMessageMap[code] = message
+        }
+    }
+    
+    public func generalErrorHandler(_ error: NSError?, serverError: NSError?, errorData: Data?) {
+        let combiner = fzip(someOrNil(with: errorToString), someOrNil(with: serverErrorToString), someOrNil(with: errorDataToString)) >>> { [$0.0, $0.1, $0.2] }
+        switch reduce(arr: combiner((error, serverError, errorData))) {
+        case .some(let arr):
+            if should(.print) {
+                print(arr)
+            }
+            if should(.debug) || should(.production) {
+                let errorString = arr.reduce("", { $0 + "\n\($1)"})
+                errorString |> (("Errors" >|> alert) >>> presenter)
+            }
+        case .none:
+            break
+        }
+    }
+    
+    public func errorHandler(_ error: NSError?) {
+        if should(.print) {
+            error ?> errorToString >>> { print($0) }
+        }
+        if should(.debug) || should(.production) {
+            error ?> (errorToString >>> ("Error" >|> alert) >>> presenter)
+        }
+    }
+    
+    public func errorDataHandler(_ errorData: Data?) {
+        if should(.print) {
+            errorData ?> errorDataToString >>> { print($0) }
+        }
+        if should(.debug) || should(.production) {
+            errorData ?> (errorDataToString >>> ("Error" >|> alert(_:_:)) >>> presenter)
+        }
+    }
+    
+    public func serverErrorHandler(_ serverError: NSError?) {
+        if should(.print) {
+            serverError ?> serverErrorToString >>> { print($0) }
+        }
+        if should(.debug) || should(.production) {
+            serverError ?> (serverErrorToString >>> ("Server Error" >|> alert) >>> presenter)
+        }
+    }
+    
+    public func errorToString(_ error: NSError) -> String {
+        return should(.debug) ? "Response Error:\n Domain: \(error.domain), Description: \(error.localizedDescription), Code: \(error.code)" : errorMessageMap[error.code] ?? "Response Error: Status code was not handled"
+    }
+    
+    public func serverErrorToString(_ serverError: NSError) -> String {
+        return should(.debug) ?  "Server Error:\n Domain: \(serverError.domain), Description: \(serverError.localizedDescription), Code: \(serverError.code)" : serverErrorMessageMap[serverError.code] ?? "Server Error: Status code was not handled."
+    }
+    
+    public func should(_ op: ErrorHandlingConfig) -> Bool {
+        return (op.rawValue & config) != 0
+    }
+}
+
+public enum SomeOrNil<T> {
```

Comment:
This feels suspiciously like `Optional<[T]>`... what is the difference?

Reply:
i'm not sure what you mean by Optional<[T]>, i just tried to the accumulating effect in the pointfree stuff and tried to fit in in the context of the error handling... there very well could be no difference

Reply:
Ah, looking at the documentation it does the same as optional, so i'll refactor to use that? If you think that's the direction worth going

Reply:
I think so... not sure, but I gotta say, that does seem like a really cool use of optional in this case